kv/index.go

```diff
@@ -0,0 +1,537 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+)
+
+const defaultPopulateBatchSize = 100
+
+// Index is used to define and manage an index for a source bucket.
+//
+// When using the index you must provide it with an IndexSource.
+// The IndexSource provides the index with the contract it needs to populate
+// the entire index and traverse a populated index correctly.
+// The IndexSource provides a way to retrieve the key on which to index with
+// when provided with the value from the source.
+// It also provides the way to access the source bucket.
+//
+// The following is an illustration of its use:
+//
+//  byUserID := func(v []byte) ([]byte, error) {
+//      auth := &influxdb.Authorization{}
+//
+//      if err := json.Unmarshal(v, auth); err != nil {
+//          return err
+//      }
+//
+//      return auth.UserID.Encode()
+//  }
+//
+//  // configure a write only index
+//  indexByUser := NewIndex(NewSource([]byte(`authorizationsbyuserv1/), byUserID), false)
```

Comment:
note to self: remove last argument `, false` as this is now default behaviour overridden by an option.